actors/builtin/market/market_state_test.go

```diff
@@ -38,4 +62,300 @@ func TestConstruction(t *testing.T) {
+			curEpoch := abi.ChainEpoch(104)
+			rt.SetEpoch(curEpoch)
+
+			assertBalances(t, rt, st, client, 550, 500)
+			assertBalances(t, rt, st, provider, 7, 7)
+
+			st.updatePendingDealState(rt, dealId, curEpoch)
+
+			assertBalances(t, rt, st, client, 540, 490)
+			assertBalances(t, rt, st, provider, 17, 7)
+
+			dealState = st.mustGetDealState(rt, dealId)
+			assert.Equal(t, curEpoch, dealState.LastUpdatedEpoch)
+		})
+
+		t.Run("provider gets partial payment after slashing", func(t *testing.T) {
+			// Set up an elapsed update time of 2 epochs
+			dealState := &DealState{
+				SectorStartEpoch: dealStartEpoch,
+				LastUpdatedEpoch: epochUndefined,
+				SlashEpoch:       abi.ChainEpoch(104),
+			}
+
+			rt, st, dealId := setup(dealState)
+			curEpoch := abi.ChainEpoch(105)
+			rt.SetEpoch(curEpoch)
+
+			assertBalances(t, rt, st, client, 550, 500)
+
+			// TODO: Find out about "payment remaining" calculation, is it correct
+			// this test fails if we don't add one more StoragePricePerEpoch to LockedBalance
+			st.AddLockedBalance(rt, client, abi.NewTokenAmount(5))
+			assertBalances(t, rt, st, client, 550, 505)
+
+			assertBalances(t, rt, st, provider, 7, 7)
+
+			st.updatePendingDealState(rt, dealId, curEpoch)
+
+			assertBalances(t, rt, st, client, 530, 0)
+			assertBalances(t, rt, st, provider, 20, 0)
+
+			rt.ExpectAbort(exitcode.ErrIllegalState, func() {
+				st.mustGetDeal(rt, dealId)
+			})
+		})
+
+		t.Run("provider slashed for failing to start the deal on time", func(t *testing.T) {
+			dealState := &DealState{
+				SectorStartEpoch: epochUndefined,
+				LastUpdatedEpoch: epochUndefined,
+				SlashEpoch:       epochUndefined,
+			}
+
+			rt, st, dealId := setup(dealState)
+			curEpoch := abi.ChainEpoch(105)
+			rt.SetEpoch(curEpoch)
+
+			//assertBalances(t, rt, st, client, 550, 500)
+			assertBalances(t, rt, st, client, 550, 500)
+			assertBalances(t, rt, st, provider, 7, 7)
+
+			st.updatePendingDealState(rt, dealId, curEpoch)
+
+			assertBalances(t, rt, st, client, 550, 0)
+			assertBalances(t, rt, st, provider, 0, 0)
+
+			rt.ExpectAbort(exitcode.ErrIllegalState, func() {
+				st.mustGetDeal(rt, dealId)
+			})
+		})
+
+		t.Run("deal is expired/completed so funds are unlocked after payment", func(t *testing.T) {
+			dealState := &DealState{
+				SectorStartEpoch: dealStartEpoch,
+				LastUpdatedEpoch: epochUndefined,
+				SlashEpoch:       epochUndefined,
+			}
+
+			rt, st, dealId := setup(dealState)
+			curEpoch := dealEndEpoch + 1
+			rt.SetEpoch(curEpoch)
+
+			assertBalances(t, rt, st, client, 550, 500)
+			assertBalances(t, rt, st, provider, 7, 7)
+
+			st.updatePendingDealState(rt, dealId, curEpoch)
+
+			// There should be no balance changes, as payments are done
+			assertBalances(t, rt, st, client, 50, 0)
+			assertBalances(t, rt, st, provider, 507, 0)
+
+			// State stops tracking this deal
+			rt.ExpectAbort(exitcode.ErrIllegalState, func() {
+				st.mustGetDeal(rt, dealId)
+			})
+		})
+	})
```

Comment:
Most of these direct-state tests are too low level. There is a bunch of technical debt in the market actor state. These tests will cement that debt and make the fixes painful. Two particular examples are:
- direct use of the runtime
- unscalable iteration over all deals in updatePendingDealState

The scenarios described in the test descriptions are great, but please exercise them via the actor's exported API, not the state object. On balance, I'd prefer to withdraw this file from the PR so the rest of it can land, then bring these scenarios back a different way.